Revise PodAutoscaler's life cycle

hohaichi opened the pull request:
Currently PodAutoscaler's Ready condition is the same as Active condition. It is not only mismatched with Revision's Ready condition, where the Active condition is only an informative condition that does not define the Ready condition, but also not capturing the PodAutoscaler's status very well. For example, when everything is good but there's no traffic, the PodAutoscaler should be Ready, but Inactive.

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes #3456 

## Proposed Changes

* Make Active an informational condition for PodAutoscaler.
* Define Bootstrap as a living condition for PodAutoscaler's Ready.
* Add ReadyReplicas to PodScalable's Status so that PodScalable can be used to detect when the PA may be stuck from scaling up from zero.
* PodAutoscaler checks the following conditions when PA got stuck not able to scale up from zero: container exiting, pod unscheduled, and image pull problems.
* Revision reads PodAutoscaler's Ready condition and updates its ContainerHealthy condition.
* PodAutoscaler scales down (to 0 if allowed) when it fails to become Ready.

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Revise PodAutoscaler's life cycle to capture its Ready status more accurately.
```

hohaichi commented:
/hold
Prototyping to gather feedback

hohaichi commented:
/assign @mattmoor 
Matt, what do you think about this approach? I think it is extensible to fix #3077 too.

mattmoor commented on pkg/reconciler/autoscaling/kpa/kpa.go:
```diff
@@ -443,6 +472,18 @@ func reportMetrics(pa *pav1alpha1.PodAutoscaler, want int32, got int) error {
 }
 
+func (c *Reconciler) verifyContainer(pa *pav1alpha1.PodAutoscaler) (*apis.Condition, error) {
+	rev, err := c.revLister.Revisions(pa.Namespace).Get(pa.Name)
```
I don't like the layering violation that's here.  Revision creates the PodAutoscaler abstraction, so this is a back-edge in the layering graph, which I've been trying to eliminate. :(

hohaichi commented:
@mattmoor I've prototyped a fix based on our discussion.  Could you please take a look?  If the change is good, I'll add tests.

hohaichi commented:
/assign @jonjohnsonjr 
Jon, @mattmoor mentioned that you are working on separation of concerns in knative serving and doing something similar to this PR.  Could you please have a look?

vagababov commented on pkg/apis/autoscaling/v1alpha1/pa_lifecycle.go:
```diff
@@ -127,6 +127,12 @@ func (pas *PodAutoscalerStatus) IsReady() bool {
 }
 
+// IsUnready returns true if the Ready condition is false.
+func (pas *PodAutoscalerStatus) IsUnready() bool {
```
Why not just use !IsHappy()?

hohaichi replied:
Because Ready has 3 states, !IsReady() != IsUnready().

vagababov commented on pkg/apis/autoscaling/v1alpha1/pa_lifecycle.go:
```diff
@@ -179,6 +191,18 @@ func (pas *PodAutoscalerStatus) MarkResourceFailedCreation(kind, name string) {
+// MarkContainerExiting changes the "ContainerHealthy" condition to false and records the exit code.
+func (pas *PodAutoscalerStatus) MarkContainerExiting(exitCode int32, message string) {
+	exitCodeString := fmt.Sprintf("ExitCode%d", exitCode)
+	message = fmt.Sprintf("Container failed with: %s", message)
```
just `"Container failed with: " + message` -- no reason to waste cycles on formatting.

hohaichi replied:
How much difference is it?  I don't want to make the style inconsistent just for a few cycles.

vagababov commented on pkg/apis/autoscaling/v1alpha1/pa_lifecycle.go:
```diff
@@ -140,6 +146,12 @@ func (pas *PodAutoscalerStatus) IsInactive() bool {
 }
 
+// IsActive returns true if the pod autoscaler is Active.
+func (pas *PodAutoscalerStatus) IsActive() bool {
```
New methods deserve UTs.

hohaichi replied:
no tests yet.  I'd like to see the high-level idea works first.

hohaichi replied:
Tests added.

vagababov commented on pkg/apis/autoscaling/v1alpha1/pa_types.go:
```diff
@@ -99,8 +99,10 @@ const (
 	// runtime resources, and becomes true when those resources are ready.
 	PodAutoscalerConditionReady = apis.ConditionReady
-	// PodAutoscalerConditionActive is set when the PodAutoscaler's ScaleTargetRef is receiving traffic.
+	// PodAutoscalerConditionActive turns true when the PodAutoscaler's ScaleTargetRef can receive traffic.
```
Why this change?
```suggestion
	// PodAutoscalerConditionActive becomes true when the PodAutoscaler's ScaleTargetRef can receive traffic.
```

hohaichi replied:
changed "turns" to "becomes" as you suggested.  The change is because the old one isn't accurate.    A PA is active without traffic too.

vagababov commented on pkg/apis/autoscaling/v1alpha1/pa_types.go:
```diff
@@ -99,8 +99,10 @@ const (
-	// PodAutoscalerConditionActive is set when the PodAutoscaler's ScaleTargetRef is receiving traffic.
+	// PodAutoscalerConditionActive turns true when the PodAutoscaler's ScaleTargetRef can receive traffic.
 	PodAutoscalerConditionActive apis.ConditionType = "Active"
+	// PodAutoscalerConditionContainerHealthy turns true when the containers are up and running.
```
```suggestion
	// PodAutoscalerConditionContainerHealthy becomes true when the containers are up and running.
```

hohaichi replied:
done.

vagababov commented on pkg/reconciler/autoscaling/kpa/kpa.go:
```diff
@@ -423,6 +434,37 @@ func (c *Reconciler) reconcileMetric(ctx context.Context, pa *pav1alpha1.PodAuto
 	return nil
 }
 
+func (c *Reconciler) checkContainerHealth(ctx context.Context, pa *pav1alpha1.PodAutoscaler) error {
```
I expect `checkX` have no side effects. Yours actually changes PA status, so should be named differently.

hohaichi replied:
renamed to "updateContainerHealth".

jonjohnsonjr commented on pkg/reconciler/autoscaling/kpa/kpa.go:
```diff
@@ -423,6 +434,37 @@ func (c *Reconciler) reconcileMetric(ctx context.Context, pa *pav1alpha1.PodAuto
 	return nil
 }
 
+func (c *Reconciler) checkContainerHealth(ctx context.Context, pa *pav1alpha1.PodAutoscaler) error {
```
We'll want to rebase to pick up https://github.com/knative/serving/pull/4191 before moving this over to the pod autoscaler.

hohaichi replied:
rebased.

jonjohnsonjr commented on pkg/reconciler/autoscaling/kpa/kpa.go:
```diff
@@ -423,6 +434,37 @@ func (c *Reconciler) reconcileMetric(ctx context.Context, pa *pav1alpha1.PodAuto
+			logger.Errorf("Error getting pods: %v", err)
+			return err
+		} else if len(pods.Items) > 0 {
+			// Arbitrarily grab the very first pod, as they all should be crashing if one does.
```
I'm not sure if this holds -- trying to wrap my head around this:

In the revision reconciler, this was inside a check:
```go
// If a container keeps crashing (no active pods in the deployment although we want some)	
if *deployment.Spec.Replicas > 0 && deployment.Status.AvailableReplicas == 0 {
  ...
}
```

Does `pa.Status.IsActivating` imply that check is true?

hohaichi replied:
It was equivalent when I drafted this prototype, because pa.Status.Active becomes true as soon as one pod is available.  But now it is not, because pa.Status.Active only becomes true with a minimum number of replicas.  Brought that condition over.

jonjohnsonjr commented on pkg/reconciler/revision/reconcile_resources.go:
```diff
@@ -66,36 +66,6 @@ func (c *Reconciler) reconcileDeployment(ctx context.Context, rev *v1alpha1.Revi
-			// Update the revision status if pod cannot be scheduled(possibly resource constraints)
-			// If pod cannot be scheduled then we expect the container status to be empty.
-			for _, cond := range pod.Status.Conditions {
-				if cond.Type == corev1.PodScheduled && cond.Status == corev1.ConditionFalse {
```
Looks like this check got dropped in the rebase.

hohaichi replied:
Yes, this check was introduced after my prototype.  I wanted to move ContainerHealthy check from revision to PA first, then moving ResourceAvailable condition as well.  I can go ahead and move ResourceAvailable check to PA now too, if that makes sense?

hohaichi replied:
Moved this check to PodAutoscaler, with a new condition PodsHealthy.

hohaichi commented:
@mattmoor @vagababov @jonjohnsonjr 
Since all feedback are on details, it looks like the high level approach is OK with you.  I've addressed the feedback and am working on tests now.

jonjohnsonjr commented on pkg/reconciler/autoscaling/kpa/kpa.go:
```diff
@@ -209,6 +217,55 @@ func (c *Reconciler) reconcileDecider(ctx context.Context, pa *pav1alpha1.PodAut
+					if t := status.LastTerminationState.Terminated; t != nil {
+						logger.Infof("%s marking %s exiting with: %d/%s", pa.Name, status.Name, t.ExitCode, t.Message)
+						pa.Status.MarkContainerExiting(t.ExitCode, t.Message)
+					} else if w := status.State.Waiting; w != nil {
```
Sorry for stepping on your toes with this :smile: https://github.com/knative/serving/pull/4195

I ended up adding the `hasDeploymentTimedOut` check because because being in a "Waiting" state isn't actually fatal. It's only a problem if the container is waiting indefinitely.

There was some code to check if a deployment was failing to progress so we could surface an error, but in the case of image pull errors, that check would always fail because of `IsActivationRequired`:
https://github.com/knative/serving/blob/ffeb54ad9c5029e3a04e49961900c8f4a55f06e1/pkg/reconciler/revision/reconcile_resources.go#L102

I'm wondering if we could update that code based on your changes to the pod autoscaler status to actually make sense. As it is, `MarkProgressDeadlineExceeded` will never be called if the image can't be pulled :/

hohaichi replied:
Thank you for pointing it out.  PA doesn't depend on `Deployment`, and I'm trying to translate the conditions to `PodScalable`, so that not to add a `Deployment` in PA.  Is there an equivalent check not using deployments?

hohaichi replied:
I'm using Reason to detect the problem.

hohaichi commented on pkg/apis/autoscaling/v1alpha1/pa_types.go:
```diff
@@ -99,8 +99,12 @@ const (
 	PodAutoscalerConditionActive apis.ConditionType = "Active"
+	// PodAutoscalerConditionContainersHealthy becomes true when the containers are up and running.
+	PodAutoscalerConditionContainersHealthy apis.ConditionType = "ContainersHealthy"
+	// PodAutoscalerConditionPodHealthy becomes true when the pods are up and running.
```
Thanks for catching it.  I started out with `PodHealthy` but renamed it to `PodsHealthy` to make it clear that the condition is for all pods.

hohaichi replied:
Revised to Bootstrap condtion.

hohaichi commented on pkg/apis/autoscaling/v1alpha1/pa_types.go:
```diff
@@ -99,8 +99,12 @@ const (
-	// PodAutoscalerConditionActive is set when the PodAutoscaler's ScaleTargetRef is receiving traffic.
+	// PodAutoscalerConditionActive becomes true when the PodAutoscaler's ScaleTargetRef can receive traffic.
 	PodAutoscalerConditionActive apis.ConditionType = "Active"
+	// PodAutoscalerConditionContainersHealthy becomes true when the containers are up and running.
```
Yes, I wanted to make it clear that the condition is for all containers.  `ContainerHealthy` is ambiguous in whether it becomes `true` when any container becomes healthy, or becomes `true` when all containers become healthy.  I didn't rename `ContainerHealthy` in Revision though, because I'm afraid it may break backward compatibility.

hohaichi replied:
Revised to Bootstrap condition, so no conflicting in naming anymore.

hohaichi commented:
Rebased and fixed tests in scaler_test and kpa_test.  More to come...

hohaichi commented:
/hold
(work in progress)

hohaichi commented:
/hold cancel

jonjohnsonjr commented on pkg/apis/autoscaling/v1alpha1/pa_lifecycle.go:
```diff
@@ -179,17 +191,41 @@ func (pas *PodAutoscalerStatus) MarkInactive(reason, message string) {
 // MarkResourceFailedCreation changes the "Active" condition to false to reflect that a
 // critical resource of the given kind and name was unable to be created.
 func (pas *PodAutoscalerStatus) MarkResourceFailedCreation(kind, name string) {
+	// TODO: This looks more like a bootstrap condition?
```
I turned the previous comment into a graph because I wanted to visualize which resources actually update their conditions based on failure to create a child resource (in red): 
![graph](https://user-images.githubusercontent.com/17863526/60216021-d7974780-981d-11e9-8072-84d05e5cd397.png)

It seems like we should be more consistent here... cc @mattmoor

hohaichi commented:
Updated per our discussion yesterday--making resource-not-owned and resource-failed-creation readiness conditions.

mattmoor commented:
I think @jonjohnsonjr is going to reopen this as another PR.  Going to close this to clear from gubernator.
